File: Webbprogrammerare-120hp/Year-1/1dv024/kr222if-examination-3/Game.cs
using System;
using System.Collections;
using System.Collections.Generic;


/// <summary>
/// <author> Mosa Kasem Rasol </author>
/// <version> 1.0.3 </version>
/// </summary>

namespace kr222if_examination_3
{
    public class Game
    {
        /// <summary>
        /// Private players, deck and dealer
        /// These members are not visible outside the class
        /// </summary>
        private readonly List<Player> players;
        private readonly Deck deck;
        private readonly Dealer dealer;
        /// <summary>
        /// Game constructor, deck, player and dealer maker
        /// </summary>
        /// <param name="allPlayer">amount of players</param>
        public Game(int allPlayer)
        {
            dealer = new Dealer();
            deck = new Deck();
            players = new List<Player>();


            for (int i = 0; i < allPlayer; i++)
            {
                players.Add(new Player("player " + i));
            }

        }
        /// <summary>
        /// Game initiater
        /// </summary>
        public void LetsPlay()
        {
            BeginGame();
        }
        /// <summary>
        /// Game Logic
        /// </summary>
        private void BeginGame()
        {
            /// <summary> first eachplayer receives a card </summary>
            foreach (var player in players)
            {
                deck.GiveCard(player);
            }
            /// <summary> PlayerPlays until he is satisfied </summary>
            foreach (var player in players)
            {
                PlayerPlays(player);
                if (player.CountScore() >= 22)
                {
                    Console.WriteLine($"{player.ToString()} Busted \n{dealer.ToString()} Dealer Wins\n");
                    ClearHands(player, dealer);
                }
                else
                {
                    DealerPlays(player);
                    if (dealer.CountScore() > 21)
                    {
                        Console.WriteLine($"{player.ToString()} Wins \n{dealer.ToString()} BUSTED!\n");
                        ClearHands(player, dealer);
                    }  else if (dealer.CountScore() < player.CountScore()) {
                        Console.WriteLine($"{player.ToString()} Wins \n{dealer.ToString()} \n");
                        ClearHands(player, dealer);
                    }  else if (dealer.CountScore() == player.CountScore()) { 
                        Console.WriteLine($"{player.ToString()} \n{dealer.ToString()} Wins \n");
                        ClearHands(player, dealer);
                    }  else if (dealer.CountScore() > player.CountScore()) {
                        Console.WriteLine($"{player.ToString()} \n{dealer.ToString()} Wins\n");
                        ClearHands(player, dealer);
                    }
                }
            }
        }
        /// <summary>
        /// Player receives new card until he is satisfied
        /// </summary>
        /// <param name="player"></param>
        private void PlayerPlays(Player player)
        {
            while(!player.IsSatisfied())
            {
                deck.GiveCard(player);
            }
        }
        /// <summary>
        /// Dealer receives new card until he scores higher than player
        /// </summary>
        /// <param name="player"></param>
        private void DealerPlays(Player player)
        {
            while(!dealer.IsSatisfied() && dealer.CountScore() < player.CountScore())
            {
                deck.GiveCard(dealer);
            }
        }
        /// <summary>
        /// Clear hands of player and dealer
        /// </summary>
        /// <param name="player"> current player </param>
        /// <param name="dealer"> the dealer </param>
                private void ClearHands(Player player, Dealer dealer)
                {
                    deck.DiscardedCards(player, dealer);
                    dealer.Clean();
                    player.Clean();
                }
            }
        }

File: Webbprogrammerare-120hp/Year-1/1dv021/kr222if-examination-2/src/Game.js
'use strict'
/**
 * @module src/kortspel
 * @author Mosa Kasem Rasol
 * @version 1.1.1
 */

const Dealer = require('./Dealer.js')
const PlayerClass = require('./Player.js')
const Deck = require('./Deck.js')

// created a shuffled deckofcards and toss it around.

let deck = new Deck()
let dealer = new Dealer()
let deckofKards = dealer.shuffle(deck.card())
/**
 * @class CardGame
 * @param {object[]} -all the used cards
 * @param {object[]} -all players
 * @param deck - contains 52 cards
 */
class CardGame {
  constructor () {
    this.players = []
    this.allTheUsedCards = []
    this.deck = deckofKards
  }
  /**
   * player play till satisfied and controls that there are enough cards.
   * @param {Object Player} player
   * @memberof CardGame
   */
  PlayerPlays (player) {
    if (this.deck.length <= 1) {
      this.deck.push(...this.allTheUsedCards)
      this.allTheUsedCards = []
      dealer.shuffle(this.deck)
    }
  }
  /**
   * @memberof CardGame
   * Dealer plays until satisfied
   * Dealer checks if deck's are empty.
   */
  DealerPlays () {
    while (dealer.isNotSatisfied()) {
      if (this.deck.length <= 1) {
        this.deck.push(...this.allTheUsedCards)
        this.allTheUsedCards = []
      }
      dealer.givecard(this.deck)
    }
    console.log(`Dealer ${dealer.toString()}`)
    this.collectCards(dealer.hand)
    dealer.emptyHand()
  }
  collectCards (hand) {
    this.allTheUsedCards.push(...hand)
  }
  playerCreater (playersAmount = 1) {
    for (let i = 0; i < playersAmount; i++) {
      let player = new PlayerClass(i)
      this.players.push(player)
    }
  }
  /**
   * @memberof CardGame
   */
  gameStart () {
    for (let i = 0; i < this.players.length; i++) {
      this.players[i].RecieveNewCard(this.deck)
      this.PlayerPlays()
      while (this.players[i].PlayerNotSatisfied()) {
        this.players[i].RecieveNewCard(this.deck)
        this.PlayerPlays()

        if (this.players[i].sum >= 22) {
          console.log(`player ${this.players[i]} Busted! \nDealer Wins!  \n`) // if dealer gets over 21, loses
        } else if (this.players[i].sum) {
          continue
        }
      }
      if (this.players[i].sum <= 21) {
        console.log(`player ${this.players[i].toString()}`)
      }
      if (this.players[i].sum === 21) {
        console.log('Player wins! \n') // if player gets 21, wins
      }
      if (this.players[i].sum < 21) {
        this.DealerPlays()
        if (dealer.result === 21) { // if dealer gets 21, wins
          console.log('Dealer Wins! \n')
        } else if (dealer.result > 21) { // if dealer over 21, player wins
          console.log('Player wins! \n')
        } else if (this.players[i].sum > dealer.result) { // if player gets higher than dealer, wins
          console.log('Player wins! \n')
        } else if (dealer.result > this.players[i].sum) { // if dealer gets higher than player, wins
          console.log('Dealer wins \n')
        } else if (this.players[i].sum > dealer.result) { // if player gets higher than dealer, wins
          console.log('Player wins! \n')
        } else if (this.players[i].sum === dealer.result) { // if dealer and player gets same result, dealer wins
          console.log(`Dealer wins! \n`)
        } else if (this.players[i].sum > dealer.result) { // if player gets higher than dealer, wins
          console.log('Player wins! \n')
        }
      }
      this.collectCards(this.players[i].hand)
      this.players[i].emptyHand()
    }
  }
}

module.exports = CardGame
